packages/context/src/context.ts

```diff
@@ -211,18 +363,16 @@ export class Context {
    * binding is found, an error will be thrown.
    *
    * @param key Binding key
+   * @param optional If set to true, the method will return `undefined` instead
+   * of throwing an error if the binding key is not found
    */
-  getBinding(key: string): Binding {
+  getBinding(key: string, optional: boolean = false): Binding | undefined {
     Binding.validateKey(key);
-    const binding = this.registry.get(key);
-    if (binding) {
-      return binding;
-    }
-
-    if (this._parent) {
-      return this._parent.getBinding(key);
-    }
-
+    const result = this.visitAllContexts(ctx => {
+      const binding = ctx.registry.get(key);
+      return {done: binding != null, value: binding};
+    });
```

Comment:
Building on top of my earlier comments about `map`, `takeUntil` and `reduce`, and assuming we have some sort of a wrapper for easier composition of iterable-processing functions, this code would be rewritten as follows:

```ts
const result = _(this.contexts())
  .map(ctx => ctx.registry.get(key)
  .takeUntil(binding => binding != null)
  .first();
```

Sadly, lodash does not support lazy iteration of potentially infinite IteratorIterable collections (see https://github.com/lodash/lodash/issues/737), otherwise we would be able to leverage their implementation out of the box.